src/node.jl

```diff
@@ -155,24 +167,99 @@ compute(::Node)::Nothing =
 setupIO!(::Node, ::NodeIO) =
     error("The `setupIO` method hasn't been implemented for your node yet!")
 
-# These methods can be overwritten as needed
+##############################
+# OPTIONAL INTERFACE
+##############################
 startup(::Node)::Nothing = nothing
 getcontext(node::Node)::Union{Nothing,ZMQ.Context} = getIO(node).ctx
-getrate(node::Node)::Float64 = node.rate
 getIO(node::Node)::NodeIO = node.nodeio
+
+function getname(::T) where {T<:Node}
+    # Note that this only works well when each node is only instantiated once
+    return string(T)
+end
+
+##############################
+# PROVIDED INTERFACE (don't change)
+##############################
+getoptions(node::Node) = getIO(node).opts
+getflags(node::Node) = getIO(node).flags
+getrate(node::Node)::Float64 = getoptions(node).rate 
 function isnodedone(node::Node)::Bool
     nodeio = getIO(node)
     finished_sub = any(istaskdone.(nodeio.sub_tasks))
-
-    return node.should_finish || finished_sub
+    return getflags(node).should_finish[] || finished_sub
+end
+function stopnode(node::Node; timeout=1.0)
+    getflags(node).should_finish[] = true
+    t_start = time() 
+    while (time() - t_start < timeout)
+        yield()
+        if !(getflags(node).is_running[])
+            return true
+        end
+    end
+    # If it doesn't stop in time, forcefully close the node
+    @warn "Node timed out. Forcefully closing the node."
+    closeall(node)
+    return false
 end
 
-function getname(::T) where {T<:Node}
-    # Note that this only works well when each node is only instantiated once
-    return string(T)
+publishers(node::Node) = getIO(node).pubs
+subscribers(node::Node) = getIO(node).subs
+
+for pubsub in ((:publisher, :pubs), (:subscriber, :subs))
+    @eval $(Symbol("get", pubsub[1]))(node::Node, index::Integer) = getIO(node).$(pubsub[2])[index]
+    @eval function $(Symbol("get", pubsub[1]))(node::Node, name::String)
+        index = findfirst(getIO(node).$(pubsub[2])) do msg
+            getname(msg) == name
+        end
+        if !isnothing(index)
+            return $(Symbol("get", pubsub[1]))(node, index)
+        end
+        return nothing
+    end
+    @eval $(Symbol("num", pubsub[1], "s"))(node::Node) = length(getIO(node).$(pubsub[2]))
```

Comment:
I totally incapable of parsing this chunk of code. Is this generating a list of functions?

Reply:
Haha, it's literally just defining:
```
numsubscribers(node::Node) = length(getIO(node).subs)
numpublishers(node::Node) = length(getIO(node).pubs)
```
Since there was a bunch of near-duplicate functionality between publishers and subscribers, I just wrapped it in a for-loop with `@eval` to generate both at the same time.